flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadSchedulerTest.java

```diff
@@ -192,6 +203,132 @@ public void testOnReadBufferRequestError() throws Exception {
         assertAllResourcesReleased();
     }
 
+    @Test
+    public void testBufferRequestTimeoutWithoutRecycle() throws Exception {
+        occupyBuffers();
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.createSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        readScheduler.allocateBuffers(new PriorityQueue<>(allReaders));
+
+        assertExpectedTimeoutException(subpartitionReader.getFailureCause());
+    }
+
+    @Test
+    public void testBufferRequestSuccessWithRecycle() throws Exception {
+        List<MemorySegment> buffers = occupyBuffers();
+
+        // Recycle the buffer after some time
+        Random random = new Random();
+        int recycleBufferWaitMs = random.nextInt(2999);
+        startBufferRecycleThread(bufferPool, buffers, recycleBufferWaitMs, false);
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.createSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        readScheduler.allocateBuffers(new PriorityQueue<>(allReaders));
+
+        assertNull(subpartitionReader.getFailureCause());
+    }
+
+    @Test
+    public void testBufferRequestSuccessAfterRecycle() throws Exception {
+        List<MemorySegment> buffers = occupyBuffers();
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.createSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        // Recycle the buffer after 1s
+        startBufferRecycleThread(bufferPool, buffers, 1000, false);
+
+        Thread.sleep(3500);
+
+        readScheduler.allocateBuffers(new PriorityQueue<>(allReaders));
+        assertNull(subpartitionReader.getFailureCause());
+    }
+
+    @Test
+    public void testBufferRequestTimeoutAfterRecycle() throws Exception {
+        List<MemorySegment> buffers = occupyBuffers();
+
+        SortMergeSubpartitionReader subpartitionReader =
+                readScheduler.crateSubpartitionReader(
+                        new NoOpBufferAvailablityListener(), 0, partitionedFile);
+        Set<SortMergeSubpartitionReader> allReaders = new HashSet<>();
+        allReaders.add(subpartitionReader);
+        // Recycle the buffer after 1s
+        startBufferRecycleThread(bufferPool, buffers, 400, true);
+
+        Thread.sleep(3500);
+
+        readScheduler.allocateBuffers(new PriorityQueue<>(allReaders));
+        assertExpectedTimeoutException(subpartitionReader.getFailureCause());
+    }
+
```

Comment:
I think there is one more case we need to test: after we recycle one buffer the timeout time should be refreshed.

Reply:
Added the test case, `testRequestTimeoutIsRefreshedButFailed` and `testRequestTimeoutIsRefreshedAndSuccess`.